File: client/src/layout/computeLayout.ts
// ---------------------------------------------------------------------------
// Layout module – extracted from CallRoom.tsx
//
// Exports the original harmonic-mean grid algorithm (computeStageLayout) plus
// a new computeLayout() wrapper that accepts a CallScene and returns a
// LayoutResult with absolute tile positions.
// ---------------------------------------------------------------------------

// ===== Legacy constants (exported for backward compat during migration) =====

export const MIN_STAGE_TILE_ASPECT = 9 / 16;
export const MAX_STAGE_TILE_ASPECT = 16 / 9;
export const DEFAULT_STAGE_TILE_ASPECT = 16 / 9;
export const STAGE_TILE_GAP_PX = 12;

// ===== Legacy types (exported for backward compat during migration) =========

export type StageTileSpec = {
    cid: string;
    aspectRatio: number;
};

export type StageTileLayout = {
    cid: string;
    width: number;
    height: number;
};

export type StageRowLayout = {
    items: StageTileLayout[];
};

// ===== Legacy functions (exact copies from CallRoom.tsx) ====================

export function clampStageTileAspectRatio(ratio?: number | null): number {
    if (!ratio || !Number.isFinite(ratio) || ratio <= 0) {
        return DEFAULT_STAGE_TILE_ASPECT;
    }
    return Math.min(MAX_STAGE_TILE_ASPECT, Math.max(MIN_STAGE_TILE_ASPECT, ratio));
}

export function computeStageLayout(tiles: StageTileSpec[], availableWidth: number, availableHeight: number, gap: number): StageRowLayout[] {
    if (tiles.length === 0 || availableWidth <= 0 || availableHeight <= 0) {
        return [];
    }

    const candidateRows: number[][][] = tiles.length === 1
        ? [[[0]]]
        : tiles.length === 2
            ? [[[0, 1]], [[0], [1]]]
            : [[[0, 1, 2]], [[0, 1], [2]], [[0], [1, 2]], [[0], [1], [2]]];

    let bestLayout: StageRowLayout[] = [];
    let bestHarmonicShortEdge = -1;
    let bestMinShortEdge = -1;
    let bestArea = -1;
    let bestRowCount = Number.POSITIVE_INFINITY;

    for (const rows of candidateRows) {
        const baseHeights = rows.map((row) => {
            const totalAspect = row.reduce((sum, index) => sum + tiles[index].aspectRatio, 0);
            const rowWidth = availableWidth - gap * Math.max(0, row.length - 1);
            return rowWidth > 0 && totalAspect > 0 ? rowWidth / totalAspect : 0;
        });

        const verticalGap = gap * Math.max(0, rows.length - 1);
        const totalBaseHeight = baseHeights.reduce((sum, value) => sum + value, 0);
        if (totalBaseHeight <= 0 || availableHeight <= verticalGap) {
            continue;
        }

        const scale = Math.min(1, (availableHeight - verticalGap) / totalBaseHeight);
        if (scale <= 0) {
            continue;
        }

        const layout = rows.map((row, rowIndex) => {
            const rowHeight = Math.max(1, Math.floor(baseHeights[rowIndex] * scale));
            const items = row.map((index) => {
                const tile = tiles[index];
                return {
                    cid: tile.cid,
                    width: Math.max(1, Math.floor(tile.aspectRatio * rowHeight)),
                    height: rowHeight
                };
            });
            return { items };
        });

        const area = layout.reduce((sum, row) => (
            sum + row.items.reduce((rowArea, tile) => rowArea + tile.width * tile.height, 0)
        ), 0);
        const shortEdges = layout.flatMap((row) => row.items.map((tile) => Math.min(tile.width, tile.height)));
        const minShortEdge = shortEdges.reduce((currentMin, shortEdge) => Math.min(currentMin, shortEdge), Number.POSITIVE_INFINITY);
        const harmonicShortEdge = shortEdges.length / shortEdges.reduce((sum, shortEdge) => sum + (1 / shortEdge), 0);
        const rowCount = layout.length;

        const shortEdgeGainIsMeaningful = harmonicShortEdge > bestHarmonicShortEdge + 6;
        const shortEdgeIsComparable = Math.abs(harmonicShortEdge - bestHarmonicShortEdge) <= 6;
        const minShortEdgeImproved = minShortEdge > bestMinShortEdge + 1;
        const minShortEdgeComparable = Math.abs(minShortEdge - bestMinShortEdge) <= 1;

        if (
            shortEdgeGainIsMeaningful ||
            (
                shortEdgeIsComparable && (
                    rowCount < bestRowCount ||
                    (rowCount === bestRowCount && (
                        minShortEdgeImproved ||
                        (minShortEdgeComparable && area > bestArea)
                    ))
                )
            ) ||
            (
                !shortEdgeIsComparable &&
                harmonicShortEdge > bestHarmonicShortEdge &&
                minShortEdgeImproved
            )
        ) {
            bestHarmonicShortEdge = harmonicShortEdge;
            bestMinShortEdge = minShortEdge;
            bestArea = area;
            bestRowCount = rowCount;
            bestLayout = layout;
        }
    }

    return bestLayout;
}

// ===== New input types ======================================================

export interface CallScene {
    viewportWidth: number;
    viewportHeight: number;
    safeAreaInsets: Insets;
    participants: SceneParticipant[];
    localParticipantId: string;
    activeSpeakerId: string | null;
    pinnedParticipantId: string | null;
    contentSource: ContentSource | null;
    userPrefs: UserLayoutPrefs;
}

export interface SceneParticipant {
    id: string;
    role: 'local' | 'remote';
    videoEnabled: boolean;
    videoAspectRatio: number | null;
}

export interface ContentSource {
    type: 'screenShare' | 'worldCamera' | 'compositeCamera';
    ownerParticipantId: string;
    aspectRatio: number | null;
}

export interface UserLayoutPrefs {
    swappedLocalAndRemote: boolean;
    dominantFit: 'cover' | 'contain';
}

export interface Insets {
    top: number;
    bottom: number;
    left: number;
    right: number;
}

// ===== New output types =====================================================

export type LayoutMode = 'solo' | 'pair' | 'grid' | 'focus' | 'content';
export type FitMode = 'cover' | 'contain';

export interface LayoutResult {
    mode: LayoutMode;
    tiles: TileLayout[];
    localPip: PipLayout | null;
}

export interface TileLayout {
    id: string;
    type: 'participant' | 'contentSource';
    frame: Rect;
    fit: FitMode;
    cornerRadius: number;
    zOrder: number;
}

export interface PipLayout {
    participantId: string;
    frame: Rect;
    fit: FitMode;
    cornerRadius: number;
    anchor: 'topLeft' | 'topRight' | 'bottomLeft' | 'bottomRight';
    zOrder: number;
}

export interface Rect {
    x: number;
    y: number;
    width: number;
    height: number;
}

// ===== Device-group-based layout constants ==================================

const LAYOUT_CONSTANTS = {
    gap: { phone: 8, tablet: 12, desktop: 12 },
    outerPadding: { phone: 8, tablet: 16, desktop: 16 },
    cornerRadius: { phone: 12, tablet: 14, desktop: 16 },
    minTileAspect: 9 / 16,
    maxTileAspect: 16 / 9,
    defaultTileAspect: 16 / 9,
    pipWidthFraction: { phone: 0.24, tablet: 0.20, desktop: 0.16 },
    pipAspectRatio: 3 / 4,
    pipInset: { phone: 12, tablet: 16, desktop: 20 },
    pipCornerRadius: 12,
    primaryRatio: 0.75,
    thumbnailMinSize: { phone: 72, tablet: 96, desktop: 120 },
    harmonicShortEdgeTolerance: 6,
    minShortEdgeTolerance: 1,
};

// ===== Device group classification ==========================================

type DeviceGroup = 'phone' | 'tablet' | 'desktop';

function deviceGroup(viewportWidth: number, viewportHeight: number): DeviceGroup {
    const shortSide = Math.min(viewportWidth, viewportHeight);
    if (shortSide < 600) return 'phone';
    if (shortSide < 1024) return 'tablet';
    return 'desktop';
}

// ===== Mode derivation ======================================================

function deriveMode(scene: CallScene): LayoutMode {
    if (scene.participants.length === 1) return 'solo';
    if (scene.contentSource !== null) return 'content';
    if (scene.pinnedParticipantId !== null) return 'focus';
    if (scene.participants.length === 2) return 'pair';
    return 'grid';
}

// ===== Grid tile absolute positioning =======================================

function gridTilesToAbsolute(
    rows: StageRowLayout[],
    availableX: number,
    availableY: number,
    availableWidth: number,
    availableHeight: number,
    gap: number,
): { cid: string; frame: Rect }[] {
    const totalRowsHeight =
        rows.reduce((sum, row) => sum + row.items[0].height, 0) +
        gap * Math.max(0, rows.length - 1);
    let rowY = availableY + (availableHeight - totalRowsHeight) / 2;

    const result: { cid: string; frame: Rect }[] = [];
    for (const row of rows) {
        const rowWidth =
            row.items.reduce((sum, tile) => sum + tile.width, 0) +
            gap * Math.max(0, row.items.length - 1);
        let tileX = availableX + (availableWidth - rowWidth) / 2;
        const rowHeight = row.items[0].height;

        for (const tile of row.items) {
            result.push({
                cid: tile.cid,
                frame: { x: tileX, y: rowY, width: tile.width, height: rowHeight },
            });
            tileX += tile.width + gap;
        }
        rowY += rowHeight + gap;
    }
    return result;
}

// ===== PIP computation ======================================================

function computePip(
    participantId: string,
    viewportWidth: number,
    viewportHeight: number,
    group: DeviceGroup,
    videoAspectRatio: number | null,
): PipLayout {
    const widthFraction = LAYOUT_CONSTANTS.pipWidthFraction[group];
    const inset = LAYOUT_CONSTANTS.pipInset[group];
    const width = viewportWidth * widthFraction;
    const ar = clampStageTileAspectRatio(videoAspectRatio ?? LAYOUT_CONSTANTS.pipAspectRatio);
    const height = width / ar;

    return {
        participantId,
        frame: {
            x: viewportWidth - inset - width,
            y: viewportHeight - inset - height,
            width,
            height,
        },
        fit: 'cover',
        cornerRadius: LAYOUT_CONSTANTS.pipCornerRadius,
        anchor: 'bottomRight',
        zOrder: 1,
    };
}

// ===== Main computeLayout entry point =======================================

export function computeLayout(scene: CallScene): LayoutResult {
    const mode = deriveMode(scene);
    const group = deviceGroup(scene.viewportWidth, scene.viewportHeight);
    const gap = LAYOUT_CONSTANTS.gap[group];
    const outerPadding = LAYOUT_CONSTANTS.outerPadding[group];
    const cornerRadius = LAYOUT_CONSTANTS.cornerRadius[group];

    // Available area after safe-area insets
    const availableX = scene.safeAreaInsets.left;
    const availableY = scene.safeAreaInsets.top;
    const availableWidth =
        scene.viewportWidth - scene.safeAreaInsets.left - scene.safeAreaInsets.right;
    const availableHeight =
        scene.viewportHeight - scene.safeAreaInsets.top - scene.safeAreaInsets.bottom;

    const localParticipant = scene.participants.find(
        (p) => p.id === scene.localParticipantId,
    );

    switch (mode) {
        // ----- solo: no remote tiles, local shown as PIP -----------------------
        case 'solo': {
            return {
                mode,
                tiles: [],
                localPip: localParticipant
                    ? computePip(
                          localParticipant.id,
                          scene.viewportWidth,
                          scene.viewportHeight,
                          group,
                          localParticipant.videoAspectRatio,
                      )
                    : null,
            };
        }

        // ----- pair: one tile fills area, other as PIP --------------------------
        case 'pair': {
            const remoteParticipant = scene.participants.find(
                (p) => p.role === 'remote',
            );

            const dominant = scene.userPrefs.swappedLocalAndRemote
                ? localParticipant
                : remoteParticipant;
            const pipParticipant = scene.userPrefs.swappedLocalAndRemote
                ? remoteParticipant
                : localParticipant;

            const tiles: TileLayout[] = dominant
                ? [
                      {
                          id: dominant.id,
                          type: 'participant',
                          frame: {
                              x: availableX,
                              y: availableY,
                              width: availableWidth,
                              height: availableHeight,
                          },
                          fit: scene.userPrefs.dominantFit,
                          cornerRadius: 0,
                          zOrder: 0,
                      },
                  ]
                : [];

            const localPip = pipParticipant
                ? computePip(
                      pipParticipant.id,
                      scene.viewportWidth,
                      scene.viewportHeight,
                      group,
                      pipParticipant.videoAspectRatio,
                  )
                : null;

            return { mode, tiles, localPip };
        }

        // ----- grid / focus / content: harmonic-mean grid -----------------------
        case 'grid':
        case 'focus':
        case 'content': {
            const remoteParticipants = scene.participants.filter(
                (p) => p.role === 'remote',
            );

            const stageTiles: StageTileSpec[] = remoteParticipants.map((p) => ({
                cid: p.id,
                aspectRatio: clampStageTileAspectRatio(p.videoAspectRatio),
            }));

            // Grid available area = viewport minus safe area minus outerPadding
            const gridX = availableX + outerPadding;
            const gridY = availableY + outerPadding;
            const gridWidth = availableWidth - outerPadding * 2;
            const gridHeight = availableHeight - outerPadding * 2;

            const rows = computeStageLayout(stageTiles, gridWidth, gridHeight, gap);

            const absoluteTiles = gridTilesToAbsolute(
                rows,
                gridX,
                gridY,
                gridWidth,
                gridHeight,
                gap,
            );

            const tiles: TileLayout[] = absoluteTiles.map((t, index) => ({
                id: t.cid,
                type: 'participant' as const,
                frame: t.frame,
                fit: 'cover' as FitMode,
                cornerRadius,
                zOrder: index,
            }));

            const localPip = localParticipant
                ? computePip(
                      localParticipant.id,
                      scene.viewportWidth,
                      scene.viewportHeight,
                      group,
                      localParticipant.videoAspectRatio,
                  )
                : null;

            return { mode, tiles, localPip };
        }
    }
}

File: client-android/app/src/main/java/app/serenada/android/layout/ComputeLayout.kt
package app.serenada.android.layout

import app.serenada.android.call.ContentTypeWire
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

// ===== Input types ===========================================================

data class CallScene(
    val viewportWidth: Float,
    val viewportHeight: Float,
    val safeAreaInsets: Insets,
    val participants: List<SceneParticipant>,
    val localParticipantId: String,
    val activeSpeakerId: String?,
    val pinnedParticipantId: String?,
    val contentSource: ContentSource?,
    val userPrefs: UserLayoutPrefs,
)

data class SceneParticipant(
    val id: String,
    val role: ParticipantRole,
    val videoEnabled: Boolean,
    val videoAspectRatio: Float?,
)

enum class ParticipantRole { LOCAL, REMOTE }

data class ContentSource(
    val type: ContentType,
    val ownerParticipantId: String,
    val aspectRatio: Float?,
)

enum class ContentType {
    SCREEN_SHARE, WORLD_CAMERA, COMPOSITE_CAMERA;

    companion object {
        fun fromWire(wire: String?): ContentType = when (wire) {
            ContentTypeWire.WORLD_CAMERA -> WORLD_CAMERA
            ContentTypeWire.COMPOSITE_CAMERA -> COMPOSITE_CAMERA
            else -> SCREEN_SHARE
        }
    }
}

data class UserLayoutPrefs(
    val swappedLocalAndRemote: Boolean = false,
    val dominantFit: FitMode = FitMode.COVER,
)

data class Insets(
    val top: Float = 0f,
    val bottom: Float = 0f,
    val left: Float = 0f,
    val right: Float = 0f,
)

// ===== Output types ==========================================================

enum class LayoutMode { SOLO, PAIR, GRID, FOCUS, CONTENT }
enum class FitMode { COVER, CONTAIN }

data class LayoutResult(
    val mode: LayoutMode,
    val tiles: List<TileLayout>,
    val localPip: PipLayout?,
)

data class TileLayout(
    val id: String,
    val type: OccupantType,
    val frame: LayoutRect,
    val fit: FitMode,
    val cornerRadius: Float,
    val zOrder: Int,
)

enum class OccupantType { PARTICIPANT, CONTENT_SOURCE }

data class PipLayout(
    val participantId: String,
    val frame: LayoutRect,
    val fit: FitMode,
    val cornerRadius: Float,
    val anchor: Anchor,
    val zOrder: Int,
)

enum class Anchor { TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT }

data class LayoutRect(
    val x: Float,
    val y: Float,
    val width: Float,
    val height: Float,
)

// ===== Device-group-based layout constants ===================================

private enum class DeviceGroup { PHONE, TABLET, DESKTOP }

private object LayoutConstants {
    val gap = mapOf(DeviceGroup.PHONE to 8f, DeviceGroup.TABLET to 12f, DeviceGroup.DESKTOP to 12f)
    val outerPadding = mapOf(DeviceGroup.PHONE to 8f, DeviceGroup.TABLET to 16f, DeviceGroup.DESKTOP to 16f)
    val cornerRadius = mapOf(DeviceGroup.PHONE to 12f, DeviceGroup.TABLET to 14f, DeviceGroup.DESKTOP to 16f)
    const val minTileAspect = 9f / 16f
    const val maxTileAspect = 16f / 9f
    const val defaultTileAspect = 16f / 9f
    val pipWidthFraction = mapOf(DeviceGroup.PHONE to 0.24f, DeviceGroup.TABLET to 0.20f, DeviceGroup.DESKTOP to 0.16f)
    const val pipAspectRatio = 3f / 4f
    val pipInset = mapOf(DeviceGroup.PHONE to 12f, DeviceGroup.TABLET to 16f, DeviceGroup.DESKTOP to 20f)
    const val pipCornerRadius = 12f
    const val primaryRatio = 0.75f
    val thumbnailMinSize = mapOf(DeviceGroup.PHONE to 72f, DeviceGroup.TABLET to 96f, DeviceGroup.DESKTOP to 120f)
}

// ===== Stage layout types (used by UI for row-based grid rendering) ==========

data class StageTileSpec(
    val cid: String,
    val aspectRatio: Float,
)

data class StageTileLayout(
    val cid: String,
    val widthPx: Int,
    val heightPx: Int,
)

data class StageRowLayout(
    val items: List<StageTileLayout>,
)

// ===== Device group classification ===========================================

private fun deviceGroup(viewportWidth: Float, viewportHeight: Float): DeviceGroup {
    val shortSide = min(viewportWidth, viewportHeight)
    if (shortSide < 600f) return DeviceGroup.PHONE
    if (shortSide < 1024f) return DeviceGroup.TABLET
    return DeviceGroup.DESKTOP
}

// ===== Mode derivation =======================================================

private fun deriveMode(scene: CallScene): LayoutMode {
    if (scene.participants.size == 1) return LayoutMode.SOLO
    if (scene.contentSource != null) return LayoutMode.CONTENT
    if (scene.pinnedParticipantId != null) return LayoutMode.FOCUS
    if (scene.participants.size == 2) return LayoutMode.PAIR
    return LayoutMode.GRID
}

// ===== Aspect ratio clamping =================================================

fun clampStageTileAspectRatio(ratio: Float?): Float {
    val safeRatio = ratio ?: return LayoutConstants.defaultTileAspect
    if (!safeRatio.isFinite() || safeRatio <= 0f) return LayoutConstants.defaultTileAspect
    return safeRatio.coerceIn(LayoutConstants.minTileAspect, LayoutConstants.maxTileAspect)
}

// ===== Harmonic-mean grid algorithm (exact port from CallScreen.kt) ==========

fun computeStageLayout(
    tiles: List<StageTileSpec>,
    availableWidthPx: Float,
    availableHeightPx: Float,
    gapPx: Float,
): List<StageRowLayout> {
    if (tiles.isEmpty() || availableWidthPx <= 0f || availableHeightPx <= 0f) return emptyList()

    val candidateRows =
        when (tiles.size) {
            1 -> listOf(listOf(listOf(0)))
            2 -> listOf(listOf(listOf(0, 1)), listOf(listOf(0), listOf(1)))
            else ->
                listOf(
                    listOf(listOf(0, 1, 2)),
                    listOf(listOf(0, 1), listOf(2)),
                    listOf(listOf(0), listOf(1, 2)),
                    listOf(listOf(0), listOf(1), listOf(2)),
                )
        }

    var bestLayout = emptyList<StageRowLayout>()
    var bestHarmonicShortEdge = -1f
    var bestMinShortEdge = -1f
    var bestArea = -1f
    var bestRowCount = Int.MAX_VALUE

    candidateRows.forEach { rows ->
        val baseHeights =
            rows.map { row ->
                val totalAspect = row.sumOf { index -> tiles[index].aspectRatio.toDouble() }.toFloat()
                val rowWidth = availableWidthPx - gapPx * (row.size - 1).coerceAtLeast(0)
                if (rowWidth > 0f && totalAspect > 0f) rowWidth / totalAspect else 0f
            }
        val verticalGap = gapPx * (rows.size - 1).coerceAtLeast(0)
        val totalBaseHeight = baseHeights.sum()
        if (totalBaseHeight <= 0f || availableHeightPx <= verticalGap) return@forEach

        val scale = minOf(1f, (availableHeightPx - verticalGap) / totalBaseHeight)
        if (scale <= 0f) return@forEach

        val layout =
            rows.mapIndexed { rowIndex, row ->
                val rowHeight = maxOf(1, (baseHeights[rowIndex] * scale).toInt())
                StageRowLayout(
                    items =
                        row.map { index ->
                            val tile = tiles[index]
                            StageTileLayout(
                                cid = tile.cid,
                                widthPx = maxOf(1, (tile.aspectRatio * rowHeight).toInt()),
                                heightPx = rowHeight,
                            )
                        }
                )
            }

        val shortEdges = layout.flatMap { row -> row.items.map { minOf(it.widthPx, it.heightPx).toFloat() } }
        val minShortEdge = shortEdges.minOrNull() ?: return@forEach
        val harmonicShortEdge = shortEdges.size / shortEdges.sumOf { shortEdge -> (1f / shortEdge).toDouble() }.toFloat()
        val area = layout.sumOf { row -> row.items.sumOf { it.widthPx * it.heightPx } }.toFloat()
        val rowCount = layout.size

        val shortEdgeGainIsMeaningful = harmonicShortEdge > bestHarmonicShortEdge + 6f
        val shortEdgeIsComparable = abs(harmonicShortEdge - bestHarmonicShortEdge) <= 6f
        val minShortEdgeImproved = minShortEdge > bestMinShortEdge + 1f
        val minShortEdgeComparable = abs(minShortEdge - bestMinShortEdge) <= 1f

        if (
            shortEdgeGainIsMeaningful ||
            (
                shortEdgeIsComparable && (
                    rowCount < bestRowCount ||
                        (rowCount == bestRowCount && (
                            minShortEdgeImproved ||
                                (minShortEdgeComparable && area > bestArea)
                            ))
                    )
                ) ||
            (
                !shortEdgeIsComparable &&
                    harmonicShortEdge > bestHarmonicShortEdge &&
                    minShortEdgeImproved
                )
        ) {
            bestLayout = layout
            bestHarmonicShortEdge = harmonicShortEdge
            bestMinShortEdge = minShortEdge
            bestArea = area
            bestRowCount = rowCount
        }
    }

    return bestLayout
}

// ===== Grid tile absolute positioning ========================================

private data class AbsoluteTile(
    val cid: String,
    val frame: LayoutRect,
)

private fun gridTilesToAbsolute(
    rows: List<StageRowLayout>,
    availableX: Float,
    availableY: Float,
    availableWidth: Float,
    availableHeight: Float,
    gap: Float,
): List<AbsoluteTile> {
    if (rows.isEmpty()) return emptyList()

    val totalRowsHeight =
        rows.sumOf { row -> row.items[0].heightPx.toDouble() }.toFloat() +
            gap * max(0, rows.size - 1)
    var rowY = availableY + (availableHeight - totalRowsHeight) / 2f

    val result = mutableListOf<AbsoluteTile>()
    for (row in rows) {
        val rowWidth =
            row.items.sumOf { tile -> tile.widthPx.toDouble() }.toFloat() +
                gap * max(0, row.items.size - 1)
        var tileX = availableX + (availableWidth - rowWidth) / 2f
        val rowHeight = row.items[0].heightPx.toFloat()

        for (tile in row.items) {
            result.add(
                AbsoluteTile(
                    cid = tile.cid,
                    frame = LayoutRect(
                        x = tileX,
                        y = rowY,
                        width = tile.widthPx.toFloat(),
                        height = rowHeight,
                    ),
                )
            )
            tileX += tile.widthPx.toFloat() + gap
        }
        rowY += rowHeight + gap
    }
    return result
}

// ===== PIP computation =======================================================

private fun computePip(
    participantId: String,
    viewportWidth: Float,
    viewportHeight: Float,
    group: DeviceGroup,
    videoAspectRatio: Float?,
): PipLayout {
    val widthFraction = LayoutConstants.pipWidthFraction.getValue(group)
    val inset = LayoutConstants.pipInset.getValue(group)
    val width = viewportWidth * widthFraction
    val ar = clampStageTileAspectRatio(videoAspectRatio ?: LayoutConstants.pipAspectRatio)
    val height = width / ar

    return PipLayout(
        participantId = participantId,
        frame = LayoutRect(
            x = viewportWidth - inset - width,
            y = viewportHeight - inset - height,
            width = width,
            height = height,
        ),
        fit = FitMode.COVER,
        cornerRadius = LayoutConstants.pipCornerRadius,
        anchor = Anchor.BOTTOM_RIGHT,
        zOrder = 1,
    )
}

// ===== Focus / Content: primary + filmstrip layout ==========================

private fun computePrimaryWithFilmstrip(
    primaryId: String,
    primaryType: OccupantType,
    primaryFit: FitMode,
    filmstripParticipants: List<SceneParticipant>,
    areaX: Float,
    areaY: Float,
    areaWidth: Float,
    areaHeight: Float,
    gap: Float,
    cornerRadius: Float,
    group: DeviceGroup,
): List<TileLayout> {
    val isLandscape = areaWidth > areaHeight
    val thumbnailMin = LayoutConstants.thumbnailMinSize.getValue(group)
    val filmstripCount = filmstripParticipants.size

    var primaryRatio = LayoutConstants.primaryRatio
    if (filmstripCount > 0) {
        if (isLandscape) {
            val secondaryWidth = areaWidth * (1f - primaryRatio)
            val tileHeight = (areaHeight - gap * max(0, filmstripCount - 1)) / filmstripCount
            if (tileHeight < thumbnailMin || secondaryWidth < thumbnailMin) {
                val ratioFromHeight = if (thumbnailMin * filmstripCount + gap * max(0, filmstripCount - 1) > areaHeight) 0.5f else primaryRatio
                val ratioFromWidth = 1f - thumbnailMin / areaWidth
                primaryRatio = max(0.5f, min(primaryRatio, min(ratioFromHeight, ratioFromWidth)))
            }
        } else {
            val secondaryHeight = areaHeight * (1f - primaryRatio)
            val tileWidth = (areaWidth - gap * max(0, filmstripCount - 1)) / filmstripCount
            if (tileWidth < thumbnailMin || secondaryHeight < thumbnailMin) {
                val ratioFromWidth = if (thumbnailMin * filmstripCount + gap * max(0, filmstripCount - 1) > areaWidth) 0.5f else primaryRatio
                val ratioFromHeight = 1f - thumbnailMin / areaHeight
                primaryRatio = max(0.5f, min(primaryRatio, min(ratioFromWidth, ratioFromHeight)))
            }
        }
    }

    val tiles = mutableListOf<TileLayout>()

    if (isLandscape) {
        val primaryWidth = areaWidth * primaryRatio - gap / 2f
        val secondaryWidth = areaWidth - primaryWidth - gap

        tiles += TileLayout(
            id = primaryId,
            type = primaryType,
            frame = LayoutRect(x = areaX, y = areaY, width = primaryWidth, height = areaHeight),
            fit = primaryFit,
            cornerRadius = cornerRadius,
            zOrder = 0,
        )

        if (filmstripCount > 0) {
            val stripX = areaX + primaryWidth + gap
            val tileHeight = (areaHeight - gap * max(0, filmstripCount - 1)) / filmstripCount
            filmstripParticipants.forEachIndexed { i, p ->
                tiles += TileLayout(
                    id = p.id,
                    type = OccupantType.PARTICIPANT,
                    frame = LayoutRect(
                        x = stripX,
                        y = areaY + i * (tileHeight + gap),
                        width = secondaryWidth,
                        height = tileHeight,
                    ),
                    fit = FitMode.COVER,
                    cornerRadius = cornerRadius,
                    zOrder = i + 1,
                )
            }
        }
    } else {
        val primaryHeight = areaHeight * primaryRatio - gap / 2f
        val secondaryHeight = areaHeight - primaryHeight - gap

        tiles += TileLayout(
            id = primaryId,
            type = primaryType,
            frame = LayoutRect(x = areaX, y = areaY, width = areaWidth, height = primaryHeight),
            fit = primaryFit,
            cornerRadius = cornerRadius,
            zOrder = 0,
        )

        if (filmstripCount > 0) {
            val stripY = areaY + primaryHeight + gap
            val tileWidth = (areaWidth - gap * max(0, filmstripCount - 1)) / filmstripCount
            filmstripParticipants.forEachIndexed { i, p ->
                tiles += TileLayout(
                    id = p.id,
                    type = OccupantType.PARTICIPANT,
                    frame = LayoutRect(
                        x = areaX + i * (tileWidth + gap),
                        y = stripY,
                        width = tileWidth,
                        height = secondaryHeight,
                    ),
                    fit = FitMode.COVER,
                    cornerRadius = cornerRadius,
                    zOrder = i + 1,
                )
            }
        }
    }

    return tiles
}

// ===== Stable participant ordering ==========================================

private fun participantsStableOrder(
    participants: List<SceneParticipant>,
    localParticipantId: String,
): List<SceneParticipant> {
    val remotes = participants.filter { it.id != localParticipantId }
    val local = participants.filter { it.id == localParticipantId }
    return remotes + local
}

// ===== Main entry point ======================================================

fun computeLayout(scene: CallScene): LayoutResult {
    val mode = deriveMode(scene)
    val group = deviceGroup(scene.viewportWidth, scene.viewportHeight)
    val gap = LayoutConstants.gap.getValue(group)
    val outerPadding = LayoutConstants.outerPadding.getValue(group)
    val cornerRadius = LayoutConstants.cornerRadius.getValue(group)

    // Available area after safe-area insets
    val availableX = scene.safeAreaInsets.left
    val availableY = scene.safeAreaInsets.top
    val availableWidth =
        scene.viewportWidth - scene.safeAreaInsets.left - scene.safeAreaInsets.right
    val availableHeight =
        scene.viewportHeight - scene.safeAreaInsets.top - scene.safeAreaInsets.bottom

    // Padded area
    val paddedX = availableX + outerPadding
    val paddedY = availableY + outerPadding
    val paddedWidth = availableWidth - outerPadding * 2f
    val paddedHeight = availableHeight - outerPadding * 2f

    val localParticipant = scene.participants.find { it.id == scene.localParticipantId }

    return when (mode) {
        // ----- solo ---------------------------------------------------------------
        LayoutMode.SOLO -> {
            LayoutResult(
                mode = mode,
                tiles = emptyList(),
                localPip = localParticipant?.let {
                    computePip(it.id, scene.viewportWidth, scene.viewportHeight, group, it.videoAspectRatio)
                },
            )
        }

        // ----- pair ---------------------------------------------------------------
        LayoutMode.PAIR -> {
            val remoteParticipant = scene.participants.find { it.role == ParticipantRole.REMOTE }
            val dominant = if (scene.userPrefs.swappedLocalAndRemote) localParticipant else remoteParticipant
            val pipParticipant = if (scene.userPrefs.swappedLocalAndRemote) remoteParticipant else localParticipant

            val tiles = if (dominant != null) {
                listOf(TileLayout(
                    id = dominant.id, type = OccupantType.PARTICIPANT,
                    frame = LayoutRect(availableX, availableY, availableWidth, availableHeight),
                    fit = scene.userPrefs.dominantFit, cornerRadius = 0f, zOrder = 0,
                ))
            } else emptyList()

            LayoutResult(
                mode = mode, tiles = tiles,
                localPip = pipParticipant?.let {
                    computePip(it.id, scene.viewportWidth, scene.viewportHeight, group, it.videoAspectRatio)
                },
            )
        }

        // ----- grid ---------------------------------------------------------------
        LayoutMode.GRID -> {
            val remoteParticipants = scene.participants.filter { it.role == ParticipantRole.REMOTE }
            val stageTiles = remoteParticipants.map { StageTileSpec(it.id, clampStageTileAspectRatio(it.videoAspectRatio)) }
            val rows = computeStageLayout(stageTiles, paddedWidth, paddedHeight, gap)
            val absoluteTiles = gridTilesToAbsolute(rows, paddedX, paddedY, paddedWidth, paddedHeight, gap)

            LayoutResult(
                mode = mode,
                tiles = absoluteTiles.mapIndexed { index, t ->
                    TileLayout(t.cid, OccupantType.PARTICIPANT, t.frame, FitMode.COVER, cornerRadius, index)
                },
                localPip = localParticipant?.let {
                    computePip(it.id, scene.viewportWidth, scene.viewportHeight, group, it.videoAspectRatio)
                },
            )
        }

        // ----- focus --------------------------------------------------------------
        LayoutMode.FOCUS -> {
            val pinnedParticipant = scene.participants.find { it.id == scene.pinnedParticipantId }
            if (pinnedParticipant == null) {
                return computeLayout(scene.copy(pinnedParticipantId = null))
            }

            val secondaryParticipants = participantsStableOrder(
                scene.participants.filter { it.id != pinnedParticipant.id },
                scene.localParticipantId,
            )

            LayoutResult(
                mode = mode,
                tiles = computePrimaryWithFilmstrip(
                    pinnedParticipant.id, OccupantType.PARTICIPANT, scene.userPrefs.dominantFit,
                    secondaryParticipants, paddedX, paddedY, paddedWidth, paddedHeight,
                    gap, cornerRadius, group,
                ),
                localPip = null,
            )
        }

        // ----- content ------------------------------------------------------------
        LayoutMode.CONTENT -> {
            val cs = scene.contentSource
            if (cs == null) {
                return computeLayout(scene.copy(contentSource = null))
            }

            val secondaryParticipants = participantsStableOrder(
                scene.participants.filter { it.id != cs.ownerParticipantId },
                scene.localParticipantId,
            )

            LayoutResult(
                mode = mode,
                tiles = computePrimaryWithFilmstrip(
                    cs.ownerParticipantId + "_content", OccupantType.CONTENT_SOURCE, scene.userPrefs.dominantFit,
                    secondaryParticipants, paddedX, paddedY, paddedWidth, paddedHeight,
                    gap, cornerRadius, group,
                ),
                localPip = null,
            )
        }
    }
}
